src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -205,6 +211,38 @@ private void ChangeProjectProperty<T>(ref T field, T newValue, Func<Solution, So
             }
         }
 
+        /// <summary>
+        /// Reports a telemetry event if compilation information is being thrown away after being previously computed
+        /// </summary>
+        private static void TryReportCompilationThrownAway(SolutionState solutionState, ProjectId projectId)
+        {
+            // We log the number of syntax trees that have been parsed even if there was no compilation created yet
```

Comment:
❔ Why are we gathering this in telemetry instead of just capturing the information as part of ETL traces submitted with performance reports? See https://github.com/dotnet/roslyn/pull/38194 for an example of a related case.

Reply:
I guess I don't know. I don't know what a "performance report" is. Is that something @panopticoncentral can read? If there is already a system in place for collecting this information then that's great, and I guess its a matter of telling Paul where to get that data?

My understanding was that this telemetry event would be used as part of other telemetry events that come from CPS, VS, possibly Nuget, to get a handle on what possible wasted work might be happening on solution load.

Reply:
The thought was to be able to construct models of solution opens across the telemetry, rather than only being able to do it for explicitly captured traces.

Reply:
What steps are in place to ensure this telemetry can be disabled if it is observed to produce large amounts of data and/or have an observable impact on client load times? Roslyn aims for these operations to be efficient (i.e. throwing away a Compilation is not a bug), so putting a telemetry invocation on the path seems to defeat its goal.

Reply:
One early option would disabling this telemetry if `IsFullyLoadedAsync` is true:

https://github.com/dotnet/roslyn/blob/28b4508467cdc2ad2af7cb2f52c74cc73515a7e8/src/Workspaces/Core/Portable/Workspace/Host/Status/IWorkspaceStatusService.cs#L46

This minimizes the impact of telemetry collection to a worst-case scenario of slower solution opening, while allowing us to expand the collection in the future if we both 1) demonstrate that it is not expensive and 2) demonstrate that analysis is in place such that the additional data is likely to directly result in a better future product.

Reply:
Definitely this is only needed for solution open initially, and I agree, the aim was to add a little bit of telemetry now and add more only if it proves useful. That's why this is limited to changes that don't happen often during "normal" use. I'm fine with adding an explicit check for fully-loaded-ness.

Reply:
That is annoyingly async. Is [this pattern](https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs#L166-L174) acceptable here? I would assume we don't want to block anything here.

Reply:
Yes that pattern is fine. Also can file a bug for me to just make `IsFullyLoadedAsync` a synchronous method so this is handled automatically.